Shared mem

MattJaniszewski opened the pull request:
I ran into a limit when the FIFO pipe would get full (> ~62k of data), and cause the child processes to block until the pipe was cleared (e.g., by running the `cat` command in the appropriate FIFO handle on the file system). The use of the `shmop_` functions seemed like a good alternative.

In the PR, I added a unit test which re-creates the problem that kept the original version from working on a typical Linux system.

One thing I will note is that while I did all the due-diligence I could with the `shmop_` functions, I had to end up using `@shmop_open` with the error suppressor operator twice, to test for the presence for an existing memory segment. It's not ideal, but seemed like the only way.

A decent way to test that there are no left-over open memory segments is to run `ipcs -m` on the Linux system. You can then pass any offending `shmid` value over to the `ipcrm` binary. My testing as of yet indicated no issues of this kind, but it's theoretically possible it can happen in some edge-cases.

seyfer commented:
:+1:

githoober commented:
@MattJaniszewski: :+1:

githoober commented on src/Spork/SharedMem.php:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return array An array of messages
+     */
+    public function receiveMessages()
```
In what cases will the result value contain more then one message?

MattJaniszewski replied:
Thanks Oleg. I took a look again at this code, and realized that this method wasn't actually working as advertised. I think that I fixed the logic now, so take another look when you can.

githoober commented on src/Spork/ProcessManager.php:
```diff
@@ -99,22 +99,22 @@ public function fork($callable)
 
             // phone home on shutdown
-            register_shutdown_function(function() use(&$fifo, &$message) {
+            register_shutdown_function(function () use (&$sharedMem, &$message) {
```
Is there a reason to pass two parameters by a reference?

MattJaniszewski replied:
I removed them, since it appears to be not necessary in my case anymore.

kriswallsmith commented:
Would you mind removing the PSR-2 commit (or moving it to another pull request) so that this pull request is limited to the shared memory change?

MattJaniszewski commented:
I rebased this branch, and removed all the un-related modifications. I created separate branches for them, and I'll create the pull requests later (so they will end up being fast-forward merges).
